[Guide]: Resilient search experiences (#3248)
GoogleChrome/web.dev: src/site/content/en/reliable/resilient-search-experiences/index.md

---
layout: post
title: Resilient search experiences
authors:
  - demianrenzulli
  - jeffposnick

> date: 2018-06-16

Reviewer: Is this an old doc?

Author: Sorry, I copied the header from another post, and forgot to edit the date. Correcting this now.

> date: 2020-07-19

Reviewer: ```suggestion
date: 2020-06-19
```
@demianrenzulli I'm assuming you meant to set this date to today, not a month from today ;)

description: |
  Even in locations with fast networks a user might lose connection or connect to a flaky network, at some moments of the day.

>   Use the  Background Sync API and Web Push Notifications to provide a resilient search experience in these cases.
> ---

Reviewer: ```suggestion
  Use the  Background Sync API and Web Push Notifications to provide a resilient search experience in these cases.
draft: true
---
```

> Even in locations with fast networks a user might lose connection or connect to a flaky network, at some moments of the day.
> 
> For example: a user is on the subway searching on the phone for a product on an e-commerce website. She types the product name, clicks the "search" button, and while waiting for the results, the connection is lost, leading to the standard browser offline page.

Reviewer: ```suggestion
Even in locations with fast networks a user might lose connection or connect to a flaky network, at some moments of the day.
For example: a user is on the subway searching on the phone for a product on an e-commerce website. She types the product name, clicks the "search" button, and while waiting for the results, the connection is lost, leading to the standard browser offline page.
```

As a result, unless the user decides to come back to the site later, and repeat the same task, the site might lose a potential transaction and customer.

To provide a more resilient search experience in these cases you can use the [Background Sync API](https://developers.google.com/web/updates/2015/12/background-sync), which persists failed queries so they can be retried once the connection is recovered. This technique, in combination with [Web Push Notifications](https://developers.google.com/web/fundamentals/push-notifications) lets you inform the user of the search results, allowing you to keep them engaged with your service.

## Production case

> For concrete application of this technique let's take a look at Google Search for Chrome in Android. 
> 
> When visiting the Google Search webapp and going offline, instead of showing the standard network error page, the site serves a custom offline response, but allows users to enter their search query immediately.
> 
> The page also prompts the user to opt-in for notifications, to receive a link to the search results page once the connection is recovered.

Reviewer: ```suggestion
For concrete application of this technique let's take a look at Google Search for Chrome in Android. 
When visiting the Google Search web app and going offline, instead of showing the standard network error page, the site serves a custom offline response, but allows users to enter their search query immediately.
The page also prompts the user to opt-in for notifications, to receive a link to the search results page once the connection is recovered.
```

> 
> ![A screenshot of the background retry interface in Google Search.](search-offline-screen.png)
> 

Reviewer: ```suggestion

<figure class="w-figure">
  <img src="search-offline-screen.png" 
       alt="A screenshot of the background retry interface in Google Search.">
</figure>

```

![A service worker and a cache object communicating wiith each other.](workbox-background-sync.png)

To implement a resilient search experience in Workbox, first, import the following modules in your service worker:

```javascript
import {BackgroundSyncPlugin} from 'workbox-background-sync'; 
import {registerRoute} from 'workbox-routing'; 
import {NetworkOnly} from 'workbox-strategies';
```

Next, create an instance of the [workbox.backgroundSync plugin](https://developers.google.com/web/tools/workbox/reference-docs/v4/workbox.backgroundSync.Plugin), to automatically add failed requests to a queue, so they can be retried later:

```javascript
const bgSyncPlugin = new workbox.backgroundSync.Plugin('offlineQueryQueue', {
  maxRetentionTime: 60,
  onSync: async ({queue}) => {
    let entry;
    while ((entry = await queue.shiftRequest())) {
      try {
        const response = await fetch(entry.request);
        const cache = await caches.open('offline-search-responses');

>         const offlineUrl = entry.request.url + '&notification=true';

Reviewer: ```suggestion
        const offlineUrl = `${entry.request.url}&notification=true`;
```
Any reason not to use template literal here?

Finally, define a [networkOnly](https://developers.google.com/web/tools/workbox/modules/workbox-strategies#network_only) runtime caching strategy for requests to the search URL (e.g. `/search_action`) and pass it the `bgSyncPlugin` defined previously:

```javascript
workbox.routing.registerRoute(
  matchSearchUrl,
  new workbox.strategies.NetworkOnly({
        plugins: [bgSyncPlugin]
  })
);
```

This tells Workbox to always go to the network when the service worker intercepts a request for the search endpoint, and to delegate to the Background Sync plugin the task of managing offline scenarios.

As a result, when the user goes offline while searching, the query is automatically saved. When the connection is recovered the offline logic is triggered to process the request and inform the user of the result.

## Conclusion

In this article you learned how to implement a search experience capable of responding gracefully to offline scenarios, by combining the [Background Sync API](https://developers.google.com/web/updates/2015/12/background-sync) and the [Push API](https://developer.mozilla.org/en-US/docs/Web/API/Push_API).

> We used Workbox to show how to implement this feature, because it's our recommended way of building web apps that use service workers, but the same can be achieved by writing vanilla service worker code.

Reviewer: ```suggestion
We used Workbox to show how to implement this feature, because it's Google's recommended way of building web apps that use service workers, but the same can be achieved by writing vanilla service worker code.
```
@jeffposnick can we say this or is it too much?

Reviewer 2: I don't know if we need to make the claim that "Google" recommends usage of Workbox. (Folks should use whatever makes the most sense for them.)

How about:

> We used Workbox to show how to implement this feature, as it simplifies the process, but the same can be achieved by writing vanilla service worker code.

Author: This suggestin LGTM! Let's see what @kaycebasques has to say 👍

Reviewer: Jeff's rewrite LGTM

---
layout: post
title: Resilient search experiences
authors:
  - demianrenzulli
  - jeffposnick
date: 2020-06-19
description: |
  Even in locations with fast networks a user might lose connection or connect to a flaky network, at some moments of the day.

>   Use the  Background Sync API and Web Push Notifications to provide a resilient search experience in these cases.
> draft: true

Reviewer: ```suggestion
  Use the  Background Sync API and Web Push Notifications to provide a resilient search experience in these cases.
```

> draft: true

Reviewer: ```suggestion
```
